substrate/primitives/core/src/crypto_bytes.rs

```diff
@@ -231,14 +240,140 @@ impl<T> CryptoBytes<64, T> {
 	}
 }
 
-/// Tag used for generic public key bytes.
-pub struct PublicTag;
+mod public_bytes {
+	use super::*;
+
+	/// Tag used for generic public key bytes.
+	pub struct PublicTag;
+
+	/// Generic encoded public key.
+	pub type PublicBytes<const N: usize, SubTag> = CryptoBytes<N, (PublicTag, SubTag)>;
+
+	impl<const N: usize, SubTag> Derive for PublicBytes<N, SubTag> where Self: CryptoType {}
+
+	impl<const N: usize, SubTag> Public for PublicBytes<N, SubTag> where Self: CryptoType {}
+
+	impl<const N: usize, SubTag> sp_std::fmt::Debug for PublicBytes<N, SubTag>
+	where
+		Self: CryptoType,
+	{
+		#[cfg(feature = "std")]
+		fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+			let s = self.to_ss58check();
+			write!(f, "{} ({}...)", crate::hexdisplay::HexDisplay::from(&self.as_ref()), &s[0..8])
+		}
+
+		#[cfg(not(feature = "std"))]
+		fn fmt(&self, _: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+			Ok(())
+		}
+	}
+
+	#[cfg(feature = "std")]
+	impl<const N: usize, SubTag> std::fmt::Display for PublicBytes<N, SubTag>
+	where
+		Self: CryptoType,
+	{
+		fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+			write!(f, "{}", self.to_ss58check())
+		}
+	}
+
+	#[cfg(feature = "std")]
+	impl<const N: usize, SubTag> std::str::FromStr for PublicBytes<N, SubTag>
+	where
+		Self: CryptoType,
+	{
+		type Err = crate::crypto::PublicError;
+
+		fn from_str(s: &str) -> Result<Self, Self::Err> {
+			Self::from_ss58check(s)
+		}
+	}
+
+	#[cfg(feature = "serde")]
+	impl<const N: usize, SubTag> Serialize for PublicBytes<N, SubTag>
+	where
+		Self: CryptoType,
+	{
+		fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+		where
+			S: Serializer,
+		{
+			serializer.serialize_str(&self.to_ss58check())
+		}
+	}
+
+	#[cfg(feature = "serde")]
+	impl<'de, const N: usize, SubTag> Deserialize<'de> for PublicBytes<N, SubTag>
+	where
+		Self: CryptoType,
+	{
+		fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+		where
+			D: Deserializer<'de>,
+		{
+			Self::from_ss58check(&String::deserialize(deserializer)?)
+				.map_err(|e| de::Error::custom(format!("{:?}", e)))
+		}
+	}
+}
+
+mod signature_bytes {
+	use super::*;
+
+	/// Tag used for generic signature bytes.
+	pub struct SignatureTag;
+
+	/// Generic encoded signature.
+	pub type SignatureBytes<const N: usize, SubTag> = CryptoBytes<N, (SignatureTag, SubTag)>;
+
+	impl<const N: usize, SubTag> Signature for SignatureBytes<N, SubTag> where Self: CryptoType {}
 
-/// Generic encoded public key.
-pub type PublicBytes<const N: usize, SubTag> = CryptoBytes<N, (PublicTag, SubTag)>;
+	#[cfg(feature = "serde")]
+	impl<const N: usize, SubTag> Serialize for SignatureBytes<N, SubTag>
```

Comment:
nit: maybe for signature we could use something like this:
https://github.com/paritytech/polkadot-sdk/blob/463ccb8f9028ebf5f7fc53d7e835c21f43172253/substrate/primitives/storage/src/lib.rs#L43 (alreay used widely in a codebase, sometimes also as `use impl_serde::serialize as bytes`.)
The code is here: https://github.com/paritytech/parity-common/blob/master/primitive-types/impls/serde/src/serialize.rs